### PR TITLE
Lanierd/hydra 311

### PR DESCRIPTION
**Pull request by lanierd-adsk**
Support default material switching in the maya hydra viewport through a scene index

### CODE REVIEW

File: lib/mayaHydra/hydraExtensions/mayaUtils.cpp [hunk 1]
```diff
@@ -158,4 +159,43 @@ bool IsAMayaVisibilityAttribute(const MPlug& plug, bool& outVal)
...
+MObject GetDefaultStandardSurfaceShader()
+{
+    MSelectionList list;
+    MObject        node;
+
+    // Add the shader to the selection list
+    list.add("standardSurface1");
+
+    // Get the shader as an MObject
+    list.getDependNode(0, node);
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/mayaUtils.cpp]: Nit-pick: isn't this just `GetDependNodeFromNodeName("standardSurface1", node)`?

File: lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraSceneIndex.cpp [hunk 2]
```diff
@@ -1685,4 +1634,55 @@ bool MayaHydraSceneIndex::passNormalsToHydra()
     return val;
 }
 
+VtValue MayaHydraSceneIndex::_CreateDefaultMaterialFallback()
+{
+    static const MColor kDefaultGrayColor = MColor(0.5f, 0.5f, 0.5f) * 0.8f;
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraSceneIndex.cpp:1640]: Where does 0.8f come from?

File: lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraSceneIndex.h [hunk 3]
```diff
@@ -203,10 +203,16 @@ class MAYAHYDRALIB_API MayaHydraSceneIndex : public HdRetainedSceneIndex, public
...
     bool GetPlaybackRunning() const;
 
-    void SetDefaultMaterial(bool useDefMaterial);
-
     Fvp::PrimSelections UfePathToPrimSelections(const Ufe::Path& appPath) const override;
 
+    //Sdfpath of the maya default material
+    SdfPath GetDefaultMaterialPath() const{return _mayaDefaultMaterialPath;}
+
+    bool GetDefaultMaterialAlreadyCreated() const{return _defaultMaterialCreated;}
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraSceneIndex.h]: Accessor name is a bit wordy, I like following the name of the member variable `DefaultMaterialCreated()`, or if you really want `IsDefaultMaterialAlreadyCreated()`

File: lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraSceneIndex.h [hunk 4]
```diff
@@ -252,6 +258,9 @@ class MAYAHYDRALIB_API MayaHydraSceneIndex : public HdRetainedSceneIndex, public
     /// Is using an environment variable to tell if we should pass normals to Hydra when using the render item and mesh adapters
     static bool passNormalsToHydra();
 
+    ///Create the default material from the "standardsurface1" maya material or create a fallback material if it cannot be found
```
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraSceneIndex.h]: Seen this one in a few places, name of the default material is camelCase `standardSurface1`, not `standardSurface1`.

File: lib/flowViewport/sceneIndex/fvpDefaultMaterialSceneIndex.cpp [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        return false;
+    }
+
+    // Check if it has a material which is not in the exclusion list
+    HdMaterialBindingsSchema bindings = HdMaterialBindingsSchema::GetFromParent(prim.dataSource);
+    HdMaterialBindingSchema  binding = bindings.GetMaterialBinding();
+    HdPathDataSourceHandle   bindingPathDS = binding.GetPath();
+    if (bindingPathDS) { // If a mesh prim has no material( bindingPathDS is empty), apply anyway the default material
+        const SdfPath materialPath = bindingPathDS->GetTypedValue(0.0f);
+        auto    foundExcludedMaterialPath = std::find(
```
Comment (ppt-adsk) [inline lib/flowViewport/sceneIndex/fvpDefaultMaterialSceneIndex.cpp:81]: What size is this _defaultMaterialExclusionList expected to be?  We are looping over this for each prim.  If the list size is small, it's fine, but otherwise we can consider using an std::set.

**Review by ppt-adsk — COMMENTED** [review]
Looks pretty good to me, only minor questions / comments.

File: lib/flowViewport/sceneIndex/fvpDefaultMaterialSceneIndex.cpp (hunk 5 above)
Comment (lanierd-adsk) [reply lib/flowViewport/sceneIndex/fvpDefaultMaterialSceneIndex.cpp:81]: There is only one at this time, do you think I should switch to a std::set now ?

File: lib/mayaHydra/hydraExtensions/mayaUtils.cpp (hunk 1 above)
Comment (lanierd-adsk) [reply lib/mayaHydra/hydraExtensions/mayaUtils.cpp]: Ha sorry, I missed that function, I will use it now.

File: lib/flowViewport/sceneIndex/fvpDefaultMaterialSceneIndex.cpp (hunk 5 above)
Comment (ppt-adsk) [reply lib/flowViewport/sceneIndex/fvpDefaultMaterialSceneIndex.cpp:81]: No, just add a comment mentioning this as a possibility if the defaultMaterialExclusionList becomes large.

File: lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraSceneIndex.cpp (hunk 2 above)
Comment (lanierd-adsk) [reply lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraSceneIndex.cpp:1640]: I don't know, it could be anything else, I guess I've picked it up from a maya shader, but that is arbitrary as this fallback shader should not be used.

**Review by ppt-adsk — APPROVED** [review]

**Comment by lanierd-adsk** [comment]
FYI : @debloip-adsk @lilike-adsk @vlasovi @roopavr-adsk
